File: app/utils/chatbot.py
from collections import defaultdict
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.messages import AIMessage, HumanMessage
from langchain.chains import create_retrieval_chain
from langchain.chains.combine_documents import create_stuff_documents_chain
from dotenv import load_dotenv
import chromadb
from typing import Dict, List, Tuple, Any
import os
import logging
import yaml

# Configure logger for this module
logger = logging.getLogger(__name__)

def load_prompts_config():
    """
    Load prompt configuration from prompts.yaml file

    Returns:
        dict: Dictionary containing prompt configuration
    """
    try:
        with open("prompts.yaml", "r", encoding="utf-8") as file:
            prompts = yaml.safe_load(file)
            logger.debug("Prompts configuration loaded successfully from prompts.yaml")
            return prompts
    except FileNotFoundError:
        logger.warning("prompts.yaml not found, using fallback configuration")
        # Fallback configuration if prompts.yaml doesn't exist
        return {
            "system_prompt": "You are a helpful women's health assistant.",
            "context_instruction": "Based on the provided context, answer the user's question.",
            "fallback_response": "I don't have specific information about that topic.",
            "user_greeting": "Hello! How can I help you with women's health questions today?"
        }
    except Exception as e:
        logger.error(f"Error loading prompts configuration: {str(e)}")
        raise

def build_system_prompt(context_instruction: str) -> str:
    """
    Build the complete system prompt with context instruction

    Args:
        context_instruction (str): Instructions for how to use the context

    Returns:
        str: Complete system prompt with context placeholder
    """
    return f"{context_instruction}\n\nContext from documents:\n####\n{{context}}"

# Global session storage
session_history: Dict[str, List] = {}

def get_context_retriever_chain(vectordb):
    """
    Create a context retriever chain for generating responses based on the chat history and vector database

    Parameters:
    - vectordb: Vector database used for context retrieval

    Returns:
    - retrieval_chain: Context retriever chain for generating responses
    """
    logger.debug("Creating context retriever chain...")
    load_dotenv()
    MODEL = os.getenv("MODEL")
    logger.info(f"Initializing LLM with model: {MODEL}")

    # Load dynamic prompt configuration
    prompts_config = load_prompts_config()

    # Build system prompt using the configured prompts
    system_prompt_template = f"{prompts_config['system_prompt']}\n\n{prompts_config['context_instruction']}\n\nContext from documents:\n####\n{{context}}"
    logger.info(f"PROMPT TEMPLATE: {system_prompt_template}")
    logger.info(f"Using dynamic system prompt from prompts.yaml")
    logger.debug(f"System prompt template: {system_prompt_template[:200]}...")

    llm = ChatGoogleGenerativeAI(model=MODEL, temperature=0.2)
    retriever = vectordb.as_retriever()

    prompt = ChatPromptTemplate.from_messages([
        ("system", system_prompt_template),
        MessagesPlaceholder(variable_name="chat_history"),
        ("human", "{input}")
    ])

    logger.debug("Creating document chain and retrieval chain...")
    # Create chain for generating responses and a retrieval chain
    chain = create_stuff_documents_chain(llm=llm, prompt=prompt)
    retrieval_chain = create_retrieval_chain(retriever, chain)
    logger.debug("Context retriever chain created successfully")
    return retrieval_chain

def get_response(question, chat_history, vectordb):
    """
    Generate a response to the user's question based on the chat history and vector database

    Parameters:
    - question (str): The user's question
    - chat_history (list): List of previous chat messages
    - vectordb: Vector database used for context retrieval

    Returns:
    - response: The generated response
    - context: The context associated with the response
    """
    logger.debug(f"Generating response for question: {question[:100]}...")
    logger.debug(f"Chat history length: {len(chat_history)}")

    chain = get_context_retriever_chain(vectordb)

    try:
        logger.debug("Invoking retrieval chain...")
        response = chain.invoke({"input": question, "chat_history": chat_history})
        logger.info(f"Response generated successfully. Answer length: {len(response['answer'])}, Context documents: {len(response['context'])}")
        return response["answer"], response["context"]
    except Exception as e:
        logger.error(f"Error generating response: {str(e)}")
        raise

def get_answer_with_history(session_id: str, question: str, vectordb) -> Tuple[str, Dict[str, List[int]], List]:
    """
    Generate a response to the user's question based on the session chat history and vector database

    Parameters:
    - session_id (str): The unique identifier for the user session
    - question (str): The user's question
    - vectordb: Vector database used for context retrieval

    Returns:
    - response: The generated response
    - sources: Dictionary mapping source files to page numbers
    - chat_history: The updated chat history
    """
    logger.info(f"Processing question for session {session_id}: {question[:100]}...")

    # Initialize session history if it doesn't exist
    if session_id not in session_history:
        logger.info(f"Creating new session history for session: {session_id}")
        session_history[session_id] = []
    else:
        logger.debug(f"Using existing session history for {session_id} (length: {len(session_history[session_id])})")

    # Get the current chat history for this session
    chat_history = session_history[session_id]

    try:
        # Generate response based on user's query, chat history and vectorstore
        response, context = get_response(question, chat_history, vectordb)

        # Update chat history. The model uses up to 10 previous messages to incorporate into the response
        # Limit to last 5 interactions (10 messages) to keep context manageable
        chat_history = chat_history + [HumanMessage(content=question), AIMessage(content=response)]
        original_length = len(chat_history)
        if len(chat_history) > 10:
            chat_history = chat_history[-10:]
            logger.debug(f"Trimmed chat history from {original_length} to {len(chat_history)} messages")

        # Save updated chat history
        session_history[session_id] = chat_history
        logger.debug(f"Updated session history for {session_id} (new length: {len(chat_history)})")

        # Extract source metadata
        sources = defaultdict(list)
        for doc in context:
            sources[doc.metadata['source']].append(doc.metadata['page'])

        logger.info(f"Successfully processed question for session {session_id}. Sources: {len(sources)} documents")
        return response, dict(sources), chat_history

    except Exception as e:
        logger.error(f"Error processing question for session {session_id}: {str(e)}")
        raise

def get_chat_history(session_id: str) -> List:
    """
    Get the chat history for a specific session

    Parameters:
    - session_id (str): The unique identifier for the user session

    Returns:
    - List: The chat history for this session
    """
    history = session_history.get(session_id, [])
    logger.debug(f"Retrieved chat history for session {session_id}: {len(history)} messages")
    return history

def clear_chat_history(session_id: str) -> None:
    """
    Clear the chat history for a specific session

    Parameters:
    - session_id (str): The unique identifier for the user session
    """
    if session_id in session_history:
        old_length = len(session_history[session_id])
        session_history[session_id] = []
        logger.info(f"Cleared chat history for session {session_id} ({old_length} messages removed)")
    else:
        logger.warning(f"Attempted to clear non-existent session history: {session_id}")

def get_answer_from_query(question, vectordb):
    """
    Generate a response to the user's question based on the vector database
    without requiring chat history or Streamlit components.

    Parameters:
    - question (str): The user's question
    - vectordb: Vector database used for context retrieval

    Returns:
    - str: The generated answer
    """
    logger.info(f"Processing standalone query: {question[:100]}...")

    # Create an empty chat history for this single query
    chat_history = []

    try:
        # Use the existing get_response function
        response, context = get_response(question, chat_history, vectordb)
        logger.info(f"Successfully processed standalone query. Context documents: {len(context)}")
        return response
    except Exception as e:
        logger.error(f"Error processing standalone query: {str(e)}")
        raise

if __name__ == "__main__":
    """
    Test the chatbot functionality directly from the command line
    """
    import os
    from prepare_vectordb import get_vectorstore
    
    # Fix for protobuf compatibility issue
    import os
    os.environ["PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION"] = "python"
    
    # Ensure the docs folder exists
    if not os.path.exists("docs"):
        os.makedirs("docs")
        print("Created 'docs' folder. Please add documents and run again.")
        exit(1)
        
    # Load documents from the docs folder
    upload_docs = os.listdir("docs")
    if not upload_docs:
        print("No documents found in the 'docs' folder. Please add documents before testing.")
        exit(1)
    
    print("Loading vector database...")
    vectordb = get_vectorstore(upload_docs)
    print("Vector database loaded successfully!")
    
    print("\n=== Women's Health Assistant Test Mode ===")
    print("Type 'exit' or 'quit' to end the session\n")
    
    # Create a test session ID
    test_session_id = "test_session"
    
    while True:
        question = input("Your question: ")
        if question.lower() in ["exit", "quit"]:
            break
            
        print("\nProcessing your question...\n")
        try:
            answer, sources, _ = get_answer_with_history(test_session_id, question, vectordb)
            print(f"Answer: {answer}\n")
            
            # Display sources
            if sources:
                print("Sources:")
                for source, pages in sources.items():
                    print(f"  {source}: Pages {', '.join(map(str, pages))}")
                print()
        except Exception as e:
            print(f"Error: {str(e)}\n")
    
    print("Test session ended.")

File: src/lib/chatbot.ts
import { ChatOpenAI, OpenAIEmbeddings } from '@langchain/openai';
import { ChatPromptTemplate, MessagesPlaceholder } from '@langchain/core/prompts';
import { AIMessage, HumanMessage, BaseMessage } from '@langchain/core/messages';
import { Collection } from 'chromadb';
import * as fs from 'fs/promises';
import * as yaml from 'js-yaml';
import { PROMPTS_FILE, MODEL, EMBEDDING_MODEL, OPENAI_API_KEY, OPENAI_BASE_URL, OPENAI_EMBEDDING_BASE_URL } from './config';
import { PromptConfig, ChatMessage } from './types';

// Cache prompts to avoid reading file on every request
let promptsCache: PromptConfig | null = null;
let promptsCacheTime: number = 0;
const CACHE_TTL = 5000; // 5 seconds cache
const MAX_CONTEXT_CHARS = 12000;

const embeddings = new OpenAIEmbeddings({
  openAIApiKey: OPENAI_API_KEY,
  modelName: EMBEDDING_MODEL,
  configuration: {
    baseURL: OPENAI_EMBEDDING_BASE_URL,
  },
});

const llm = new ChatOpenAI({
  apiKey: OPENAI_API_KEY,
  model: MODEL,
  temperature: 0.2,
  configuration: {
    baseURL: OPENAI_BASE_URL,
  },
});

const prompt = ChatPromptTemplate.fromMessages([
  ['system', '{system_prompt}'],
  new MessagesPlaceholder('chat_history'),
  ['human', '{input}'],
]);

const chain = prompt.pipe(llm);

/**
 * Load prompt configuration from prompts.yaml file (async with caching)
 * @returns PromptConfig object
 */
async function loadPromptsConfig(): Promise<PromptConfig> {
  // Return cached version if still valid
  const now = Date.now();
  if (promptsCache && (now - promptsCacheTime) < CACHE_TTL) {
    return promptsCache;
  }

  try {
    const fileContents = await fs.readFile(PROMPTS_FILE, 'utf8');
    const prompts = yaml.load(fileContents) as PromptConfig;
    console.log(`Prompts configuration loaded successfully from ${PROMPTS_FILE}`);

    // Update cache
    promptsCache = prompts;
    promptsCacheTime = now;

    return prompts;
  } catch {
    console.warn(`${PROMPTS_FILE} not found, using fallback configuration`);
    // Fallback configuration if prompts.yaml doesn't exist
    const fallback = {
      system_prompt: "You are a helpful women's health assistant.",
      context_instruction: "Based on the provided context, answer the user's question.",
      fallback_response: "I don't have specific information about that topic.",
      user_greeting: "Hello! How can I help you with women's health questions today?",
    };

    // Cache fallback too
    promptsCache = fallback;
    promptsCacheTime = now;

    return fallback;
  }
}

// Global session storage
const sessionHistory: Map<string, BaseMessage[]> = new Map();

/**
 * Retrieve relevant documents from ChromaDB collection
 * @param collection - ChromaDB collection
 * @param question - The user's question
 * @param nResults - Number of results to retrieve
 * @returns Array of relevant documents with metadata
 */
async function retrieveContext(
  collection: Collection,
  question: string,
  nResults = 5
): Promise<Array<{ pageContent: string; metadata: Record<string, unknown> }>> {
  console.log(`Retrieving context for question: ${question.substring(0, 100)}...`);

  // Generate embedding for the question
  const queryEmbedding = await embeddings.embedQuery(question);

  // Query the collection
  const results = await collection.query({
    queryEmbeddings: [queryEmbedding],
    nResults,
  });

  // Format results as Document objects
  const docs = [];
  if (results.documents && results.documents[0]) {
    for (let i = 0; i < results.documents[0].length; i++) {
      docs.push({
        pageContent: results.documents[0][i] || '',
        metadata: results.metadatas?.[0]?.[i] || {},
      });
    }
  }

  console.log(`Retrieved ${docs.length} relevant documents`);
  return docs;
}

function buildContextText(docs: Array<{ pageContent: string }>): string {
  let remaining = MAX_CONTEXT_CHARS;
  const parts: string[] = [];

  for (const doc of docs) {
    if (!doc.pageContent) continue;
    if (doc.pageContent.length <= remaining) {
      parts.push(doc.pageContent);
      remaining -= doc.pageContent.length;
    } else {
      parts.push(doc.pageContent.slice(0, Math.max(0, remaining)));
      break;
    }
  }

  return parts.join('\n\n');
}

/**
 * Generate a response to the user's question based on the chat history and vector database
 * @param question - The user's question
 * @param chatHistory - List of previous chat messages
 * @param collection - ChromaDB collection used for context retrieval
 * @returns Object with answer and context documents
 */
async function getResponse(
  question: string,
  chatHistory: BaseMessage[],
  collection: Collection
): Promise<{ answer: string; context: Array<{ pageContent: string; metadata: Record<string, unknown> }> }> {
  console.log(`Generating response for question: ${question.substring(0, 100)}...`);
  console.log(`Chat history length: ${chatHistory.length}`);

  // Load dynamic prompt configuration (async with caching)
  const promptsConfig = await loadPromptsConfig();

  // Retrieve relevant context from vector database
  const contextDocs = await retrieveContext(collection, question);

  // Build context string from documents
  const contextText = buildContextText(contextDocs);

  // Build system prompt with context
  const systemPromptTemplate = `${promptsConfig.system_prompt}\n\n${promptsConfig.context_instruction}\n\nContext from documents:\n####\n${contextText}`;
  console.log(`Using dynamic system prompt from ${PROMPTS_FILE}`);

  try {
    console.log('Generating response with LLM...');
    const response = await chain.invoke({
      input: question,
      chat_history: chatHistory,
      system_prompt: systemPromptTemplate,
    });
    const answer = response.content as string;

    console.log(
      `Response generated successfully. Answer length: ${answer.length}, Context documents: ${contextDocs.length}`
    );
    return { answer, context: contextDocs };
  } catch (error) {
    console.error('Error generating response:', error);
    throw error;
  }
}

/**
 * Generate a response to the user's question based on the session chat history and vector database
 * @param sessionId - The unique identifier for the user session
 * @param question - The user's question
 * @param collection - ChromaDB collection used for context retrieval
 * @returns Object with response, sources (file -> pages mapping), and updated chat history
 */
export async function getAnswerWithHistory(
  sessionId: string,
  question: string,
  collection: Collection
): Promise<{
  response: string;
  sources: Record<string, number[] | string>;
  chatHistory: BaseMessage[];
}> {
  console.log(`Processing question for session ${sessionId}: ${question.substring(0, 100)}...`);

  // Initialize session history if it doesn't exist
  if (!sessionHistory.has(sessionId)) {
    console.log(`Creating new session history for session: ${sessionId}`);
    sessionHistory.set(sessionId, []);
  } else {
    console.log(
      `Using existing session history for ${sessionId} (length: ${sessionHistory.get(sessionId)!.length})`
    );
  }

  // Get the current chat history for this session
  let chatHistory = sessionHistory.get(sessionId)!;

  try {
    // Generate response based on user's query, chat history and collection
    const { answer, context } = await getResponse(question, chatHistory, collection);

    // Update chat history. The model uses up to 10 previous messages to incorporate into the response
    // Limit to last 5 interactions (10 messages) to keep context manageable
    chatHistory = [...chatHistory, new HumanMessage(question), new AIMessage(answer)];
    const originalLength = chatHistory.length;

    if (chatHistory.length > 10) {
      chatHistory = chatHistory.slice(-10);
      console.log(`Trimmed chat history from ${originalLength} to ${chatHistory.length} messages`);
    }

    // Save updated chat history
    sessionHistory.set(sessionId, chatHistory);
    console.log(`Updated session history for ${sessionId} (new length: ${chatHistory.length})`);

    // Extract source metadata - handle both PDFs and URLs
    const pdfSourcesMap: Map<string, number[]> = new Map();
    const urlSourcesSet: Set<string> = new Set();

    for (const doc of context) {
      const sourcePath = doc.metadata.source as string;
      const sourceType = doc.metadata.type as string;
      const page = doc.metadata.page as number;
      const title = doc.metadata.title as string;

      if (!sourcePath) continue;

      // Check if this is a URL source
      if (sourceType === 'html_url' || sourceType === 'pdf_url') {
        // For URLs, create citation with title + URL
        const citation = title ? `${title} - ${sourcePath}` : sourcePath;
        urlSourcesSet.add(citation);
      } else {
        // For PDFs, extract only the filename from the full path for security
        // e.g., "/path/to/document.pdf" -> "document.pdf"
        const filename = sourcePath.split('/').pop() || sourcePath;

        if (!pdfSourcesMap.has(filename)) {
          pdfSourcesMap.set(filename, []);
        }
        if (typeof page === 'number' && !pdfSourcesMap.get(filename)!.includes(page)) {
          pdfSourcesMap.get(filename)!.push(page);
        }
      }
    }

    // Combine PDF and URL sources with numbered prefix
    const sources: Record<string, number[] | string> = {};
    let index = 1;

    // Add PDF sources (with page numbers)
    for (const [filename, pages] of pdfSourcesMap.entries()) {
      pages.sort((a, b) => a - b);
      sources[`${index}. ${filename}`] = pages;
      index++;
    }

    // Add URL sources (no page numbers, just citation string)
    for (const citation of Array.from(urlSourcesSet)) {
      sources[`${index}. ${citation}`] = [];
      index++;
    }

    console.log(`Successfully processed question for session ${sessionId}. Sources: ${Object.keys(sources).length} documents`);
    return { response: answer, sources, chatHistory };
  } catch (error) {
    console.error(`Error processing question for session ${sessionId}:`, error);
    throw error;
  }
}

/**
 * Get the chat history for a specific session
 * @param sessionId - The unique identifier for the user session
 * @returns Array of chat messages
 */
export function getChatHistory(sessionId: string): BaseMessage[] {
  const history = sessionHistory.get(sessionId) || [];
  console.log(`Retrieved chat history for session ${sessionId}: ${history.length} messages`);
  return history;
}

/**
 * Clear the chat history for a specific session
 * @param sessionId - The unique identifier for the user session
 */
export function clearChatHistory(sessionId: string): void {
  if (sessionHistory.has(sessionId)) {
    const oldLength = sessionHistory.get(sessionId)!.length;
    sessionHistory.delete(sessionId);
    console.log(`Cleared chat history for session ${sessionId} (${oldLength} messages removed)`);
  } else {
    console.warn(`Attempted to clear non-existent session history: ${sessionId}`);
  }
}

/**
 * Convert BaseMessage to ChatMessage format
 * @param messages - Array of LangChain BaseMessages
 * @returns Array of ChatMessage objects
 */
export function formatChatHistory(messages: BaseMessage[]): ChatMessage[] {
  return messages.map((msg) => ({
    role: msg._getType() === 'ai' ? 'ai' : 'human',
    content: msg.content as string,
  }));
}

/**
 * Format sources into a user-friendly citation string
 * @param sources - Record of filename to page numbers
 * @returns Formatted citation string
 * @example
 * formatCitations({ "document.pdf": [1, 2, 5] })
 * // Returns: "document.pdf (pages 1, 2, 5)"
 */
export function formatCitations(sources: Record<string, number[]>): string {
  if (!sources || Object.keys(sources).length === 0) {
    return 'No sources cited';
  }

  const citations = Object.entries(sources).map(([filename, pages]) => {
    if (pages.length === 0) {
      return filename;
    }
    const pageStr = pages.length === 1 ? `page ${pages[0]}` : `pages ${pages.join(', ')}`;
    return `${filename} (${pageStr})`;
  });

  return citations.join('; ');
}
